scripts/test-certificates.sh

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env bash
+set -ueo pipefail
+source $(dirname $0)/utils.sh
+
+TMPDIR=$(mktemp -d -t tmp.XXXXXXXXXX)
+IDENTS_DIR=$TMPDIR/identities
+CERTS_DIR=$TMPDIR/certificates
+# TODO: make port numbers more dynamic
+CERTS_ADDR=127.0.0.1:11000
+CERTS_ADDR_PRIV=127.0.0.1:11001
+
+# TODO: find a better way
+kill_certificates_server() {
+  killall "certificates"
+}
+
+cleanup() {
+  if [[ -n $(ps | grep "certificates") ]]; then
+    kill_certificates_server
+  fi
+  rm -rf "$TMPDIR"
+  echo "cleaned up test successfully"
+}
+
+trap cleanup EXIT INT
+
+_certificates() {
+  subcommand=$1
+  shift
+
+  ident_dir="${IDENTS_DIR}/certificates"
+  ca_cert_path="${ident_dir}/ca.cert"
+  ca_key_path="${ident_dir}/ca.key"
+  rev_dburl="bolt://${CERTS_DIR}/revocations.db"
+
+  # NB: `--identity-dir` and `--config-dir` flags are only bound globally to subcommands
+  certificates --identity-dir "$ident_dir" \
+               --config-dir "$CERTS_DIR" \
+               "$subcommand" \
+               --signer.ca.cert-path "$ca_cert_path" \
+               --signer.ca.key-path "$ca_key_path" \
+               --server.address "$CERTS_ADDR" \
+               --server.private-address "$CERTS_ADDR_PRIV" \
+               --server.revocation-dburl="$rev_dburl" \
+               --log.level warn \
+                "$@"
+}
+
+_identity() {
+  subcommand=$1
+  rev_dburl="bolt://${IDENTS_DIR}/revocations.db"
+  shift
+
+  # NB: `--identity-dir` and `--config-dir` flags are only bound globally to subcommands
+  identity --identity-dir "$IDENTS_DIR" \
+           "$subcommand" \
+           --signer.tls.revocation-dburl "$rev_dburl" \
+           --log.level info \
+           "$@"
+}
+
+_identity_create() {
+  _identity create  $1 --difficulty 0 --concurrency 1 >/dev/null
+}
+
+_identity_create 'certificates'
+_certificates setup
+
+for i in {0..4}; do
+  email="testuser${i}@mail.example"
+  ident_name="testidentity${i}"
+
+  _identity_create $ident_name
+
+  if [[ i -gt 0 ]]; then
+    _certificates auth create "$i" "$email"
+  fi
+done
+
+exported_auths=$(_certificates auth export)
+_certificates run --signer.min-difficulty 0 &
```

Comment:
you can get the pid of this server by using `PID=$!` and then you can use that to kill it later on instead of `killall "certificates"` on line 14.

Reply:
Yeah, I tried that but `$!` was returning a pid that was different that the pid of the process that's running when you look in `ps`.

Reply:
Simplest fix to me looks like:

1. Change `_certificates` so that it invokes the `certificates` binary using `exec`, so it replaces the current process. Then `$!` here will be what you want.
2. The downside is that you'll also have to change lines 67 and 76 to call `_certificates` in the background with `&`. Call `wait` with no arguments afterward to wait for all background processes to finish.

Reply:
Alternatively use go tests which have more control over the concurrency and tests.

Reply:
Using Go would be great also, but: we'd want the Go program to invoke `certificates` as a separate process in order to be sure that the binary behaves as expected. For something intended to run as a daemon, that's pretty important to test. But to do that would take away a lot of the motivation for using Go. Still, I will almost always be in favor of rewriting things away from Bash :)

Reply:
We had prototypes for Go based tests in https://github.com/storj/storj/pull/1090/files#diff-e4d02494218358e9b7ffa5f637789a58R16

Reply:
I forgot that we can test cli code in go.